In [41]:
import numpy as np
import scipy
import pandas as pd
import numpy as np
import fnmatch
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction import stop_words
from sklearn.metrics.pairwise import cosine_similarity
from nltk.corpus import stopwords
from scipy.sparse import csr_matrix
from preprocessing import *
from affichage import *
from swSets import *

# Lost, Heroes, Jericho, Prison Break

In [42]:
path = "data"
corpus = get_corpus(path)

Dans la liste:
- chaque dictionnaire represente une serie du corpus

Dans les dictionnaires:
- les cles sont les numeros des saisons
- les valeurs sont l'indice max (indMax-1 en fait) des episodes de la saisons dans le dataframe.

In [21]:
show_dict_list = getShowDictList(path)
show_dict_list

[{1: 24, 2: 48, 3: 72, 4: 85, 5: 105, 6: 125},
 {1: 24, 2: 35, 3: 61, 4: 80},
 {1: 22, 2: 29},
 {1: 22, 2: 44, 3: 57, 4: 81}]

In [101]:
#definition de l'ensemble de stopwords
nltk_sw = set(stopwords.words('english'))
sklearn_sw = set(stop_words.ENGLISH_STOP_WORDS)
stopwords_set = nltk_sw | sklearn_sw | english_contractions_tokens |sw
l_nb = [str(i) for i in range(1000000)]
l_mots = ["don", "yeah", "hey", "okay", "oh", "uh", "yes", "ok"]
for mot in l_mots :
    stopwords_set.add(mot)
for nb in l_nb:
    stopwords_set.add(nb)

In [24]:
vectorizer = TfidfVectorizer(stop_words = stopwords_set, tokenizer=lemmatizing_tokenizer_v2)
X = vectorizer.fit_transform(corpus)
tfidf_df = pd.DataFrame(X.toarray(), columns=vectorizer.get_feature_names())

In [25]:
dict_list = []
for show_dict in show_dict_list:
    show_tfidf_vect_dict = dict()
    minInd = 0
    for season, maxInd in show_dict.items():
        show_tfidf_vect_dict[season] = scipy.sparse.csr_matrix(tfidf_df[minInd:maxInd].values)
        minInd= maxInd
    dict_list.append(show_tfidf_vect_dict)

Dans la liste:
    
- chaque dictionnaire represente une serie du corpus

Dans les dictionnaires:

- les clefs sont les numeros des saisons
- les valeurs sont des matrices sparses, dont les lignes sont les vecteurs tf-idf des episodes de la saisons donnée par la clef.

In [26]:
dict_list

[{1: <24x19932 sparse matrix of type '<class 'numpy.float64'>'
  	with 10625 stored elements in Compressed Sparse Row format>,
  2: <24x19932 sparse matrix of type '<class 'numpy.float64'>'
  	with 9610 stored elements in Compressed Sparse Row format>,
  3: <24x19932 sparse matrix of type '<class 'numpy.float64'>'
  	with 10412 stored elements in Compressed Sparse Row format>,
  4: <13x19932 sparse matrix of type '<class 'numpy.float64'>'
  	with 6376 stored elements in Compressed Sparse Row format>,
  5: <20x19932 sparse matrix of type '<class 'numpy.float64'>'
  	with 10827 stored elements in Compressed Sparse Row format>,
  6: <20x19932 sparse matrix of type '<class 'numpy.float64'>'
  	with 9333 stored elements in Compressed Sparse Row format>},
 {1: <24x19932 sparse matrix of type '<class 'numpy.float64'>'
  	with 10625 stored elements in Compressed Sparse Row format>,
  2: <11x19932 sparse matrix of type '<class 'numpy.float64'>'
  	with 3812 stored elements in Compressed Sparse 

In [111]:
#similarite cosinus entre les episodes de la saisons 1 de Prison Break
#similarities_sparse = cosine_similarity(dict_list[3][1] ,dense_output=False)
#print('pairwise sparse output:\n {}\n'.format(similarities_sparse))

# Similarité cosinus entre NCIS and NCIS Los Angeles

hello i'd like to report encoding problems


<br>
Les deux NCIS là: "ISO-8859-1" ---> en fait BOF
<br>
Lost : "utf-8"
<br>
bref à voir

In [89]:
path = "data"
corpus = get_corpus(path)

In [90]:
nbSeriesCorpus(path)

2

In [92]:
show_dict_list = getShowDictList(path)
show_dict_list

[{1: 23,
  2: 46,
  3: 70,
  4: 94,
  5: 113,
  6: 138,
  7: 162,
  8: 186,
  9: 210,
  10: 234,
  11: 258,
  12: 282,
  13: 288},
 {1: 24, 2: 48, 3: 72, 4: 96, 5: 120, 6: 144, 7: 149}]

In [94]:
NCIS_texts = list(corpus[:288])
LA_texts = list(corpus[288:])

In [97]:
c = []
NCIS_str = ""
for ep in NCIS_texts:
    NCIS_str += ep+" "
c.append(NCIS_str)
LA_str = ""
for ep in LA_texts:
       LA_str += ep+" "
c.append(LA_str)

In [100]:
len(c)

2

In [103]:
vectorizer = TfidfVectorizer(stop_words = stopwords_set, tokenizer=lemmatizing_tokenizer_v2)
X = vectorizer.fit_transform(c)
tfidf_df = pd.DataFrame(X.toarray(), columns=vectorizer.get_feature_names())

In [105]:
len(tfidf_df.index)

2

In [108]:
sparse_mat = scipy.sparse.csr_matrix(tfidf_df.values)

In [110]:
similarities_sparse = cosine_similarity(sparse_mat ,dense_output=False)
print('pairwise sparse output:\n {}\n'.format(similarities_sparse))

pairwise sparse output:
   (0, 1)	0.687275815227
  (0, 0)	1.0
  (1, 1)	1.0
  (1, 0)	0.687275815227



#### La similarité cosinus entre NCIS et NCIS Los Angeles est donc : 0.687275815227